managed-pointer/cpp-20/D0567R1.md

```diff
@@ -5,222 +5,384 @@
 
-```cpp
-namespace std {
-namespace experimental {
-namespace execution {
+A `managed_ptr` can be constructed in two ways:
 
-/* executor classes */
-class <executor-class> {
-  ...
+* for constructor (1) the  `managed_ptr` takes an existing pointer and does not allocate memory.
+* for constructor (2) the  `managed_ptr` allocates using the allocator specified by the allocatorT parameter, which defaults to the allocator of the associated *executor* if one is not provided.
 
-  template <typename... ManagedPtrTN>
-  executor_future_t<<executor-class>, void> put(ManagedPtrTN...);
-  template <typename... ManagedPtrTN>
-  executor_future_t<<executor-class>, void> get(ManagedPtrTN...);
-  template <typename Predicate, typename... ManagedPtrTN>
-  executor_future_t<<executor-class>, void> then_put(Predicate pred, ManagedPtrTN...);
-  template <typename Predicate, typename... ManagedPtrTN>
-  executor_future_t<<executor-class>, void> then_get(Predicate pred, ManagedPtrTN...);
+The default constructor for `T` must not be called by the `managed_ptr` during allocation.
 
-  ...
-};
+The destructor must not perform any synchronisation operations.
 
-/* customisation points */
-template <typename Executor, typename... ManagedPtrTN>
-executor_future_t<Executor, void> put(Executor, ManagedPtrTN...);
-template <typename Executor, typename... ManagedPtrTN>
-executor_future_t<Executor, void> get(Executor, ManagedPtrTN...);
-template <typename Predicate, typename Executor, typename... ManagedPtrTN>
-executor_future_t<Executor, void> then_put(Predicate pred, Executor, ManagedPtrTN...);
-template <typename Predicate, typename Executor, typename... ManagedPtrTN>
-executor_future_t<Executor, void> then_get(Predicate pred, Executor, ManagedPtrTN...);
+#### 3.3.2. Execution context capabilities
 
-}  // namespace execution
-}  // namespace experimental
-}  // namespace std
-```
+In order to support a wide range of systems, both heterogneous and distributed, a high level absraction for the `managed_ptr` is nessesary, however not all devices will be able to support of all of a `managed_ptr`s capabilities.
+
+As the `managed_ptr` can be passed between and used on different devices it can therefore have different capabilities depending on where it is currently being used. The `managed_ptr` will likely be compiled separately or a least have different restrictions imposed on the code that is executed depending on where it s used. It is therefore important to be able to determine what capabilities the `managed_ptr` supports where it is being used.
+
+The major distinction is whether or not the `executor_type` of a `managed_ptr` is a *host* which or a *remote*. If the `executor_type` is a *host* then it supports host based capabilities; being able to perform allocation and sycnhronisation operations, if it is a *remote* it does not. Whether the `executor_type` of a `managed_ptr` is a *host* can be queried with the static constexpr bool `is_host_based`.
```

Comment:
After thinking about this further I think there is a way in which we can describe the properties of an executor with respect to data movement in a way which is fine grained and extensible.

I have updated the paper to specify traits which can be used to query whether not an executor has a particular capability, rather than grouping them into the `host` and `remote` categories:

```cpp
template<class T> constexpr bool can_put_v = can_put<T>::value;
template<class T> constexpr bool can_get_v = can_get<T>::value;
template<class T> constexpr bool can_sync_v = can_sync<T>::value;
template<class T> constexpr bool can_allocate_v = can_allocate<T>::value;
template<class T> constexpr bool can_deallocate_v = can_deallocate<T>::value;
```